File: bill_generator.js
/**
 * Canvas-based bill image generator.
 * Overlays Bitcoin address/key text and QR codes onto the bill template.
 *
 * Port of bill_generator.py — uses HTML5 Canvas instead of Pillow.
 * Zero external dependencies (uses qr_generator.js for QR codes).
 */

// ============================================================
// Template overlay positions (pixels, from bill_generator.py)
// Bill image: 1843 x 784 pixels
// ============================================================

const BILL_WIDTH = 1843;
const BILL_HEIGHT = 784;

// Fallback sweep page URL for bill QR codes (used when window.location is
// not available, e.g. Node.js / test environments).  When running in a
// browser, the QR always uses window.location so it matches the current host.
const SWEEP_PAGE_URL = 'https://ObjSal.github.io/bitcoin-gift-paper-wallet/sweep.html';

const ADDRESS_QR_BOX    = { x1: 35,   y1: 469, x2: 319,  y2: 752 };  // 284×284
const PRIVKEY_QR_BOX    = { x1: 1525, y1: 40,  x2: 1808, y2: 324 };  // 284×285
const PRIVKEY_TEXT_AREA  = { x1: 1100, y1: 2,   x2: 1808, y2: 30  };  // Orange strip
const ADDRESS_TEXT_BOX   = { x1: 348,  y1: 694, x2: 1148, y2: 751 };  // 801×58

// Banner properties
const BANNER_COLOR      = 'rgb(253, 229, 167)';
const BANNER_TEXT_COLOR  = 'rgb(0, 161, 210)';
const BANNER_LEFT_BOX   = { x1: 1082, y1: 305, x2: 1225, y2: 339 };
const BANNER_RIGHT_BOX  = { x1: 1326, y1: 305, x2: 1425, y2: 339 };


// ============================================================
// Font helpers
// ============================================================

/**
 * Find the largest font size (descending from startSize) where text fits
 * within the given width and height constraints.
 *
 * @param {CanvasRenderingContext2D} ctx
 * @param {string} text
 * @param {number} maxWidth
 * @param {number} maxHeight
 * @param {number} startSize
 * @param {string} fontFamily
 * @returns {{ font: string, width: number, height: number, size: number }}
 */
function _fitFontToBox(ctx, text, maxWidth, maxHeight, startSize, fontFamily) {
    if (startSize === undefined) startSize = 36;
    if (fontFamily === undefined) fontFamily = '"Courier New", monospace';

    for (let size = startSize; size > 5; size--) {
        const font = `${size}px ${fontFamily}`;
        ctx.font = font;
        const metrics = ctx.measureText(text);
        const tw = metrics.width;
        // Approximate text height from font size (Canvas doesn't have reliable height metrics)
        const th = size * 1.2;
        if (tw <= maxWidth && th <= maxHeight) {
            return { font, width: tw, height: th, size };
        }
    }
    const font = `6px ${fontFamily}`;
    ctx.font = font;
    const metrics = ctx.measureText(text);
    return { font, width: metrics.width, height: 6 * 1.2, size: 6 };
}


// ============================================================
// Main generator
// ============================================================

/**
 * Generate a bill image on an HTML5 Canvas.
 *
 * @param {HTMLCanvasElement} canvas - Target canvas element
 * @param {HTMLImageElement} templateImg - Pre-loaded bill template image
 * @param {string} address - Bitcoin address (bc1q... or bc1p...)
 * @param {string} privateKeyWif - WIF-encoded private key
 * @param {string} addressType - "segwit" or "taproot"
 * @param {boolean} isTweaked - If true, add "(tweaked)" labels
 * @param {string} network - "mainnet", "testnet4", or "regtest"
 */
function generateBillOnCanvas(canvas, templateImg, address, privateKeyWif, addressType, isTweaked, network) {
    if (addressType === undefined) addressType = 'taproot';
    if (isTweaked === undefined) isTweaked = false;
    if (network === undefined) network = 'mainnet';

    canvas.width = BILL_WIDTH;
    canvas.height = BILL_HEIGHT;
    const ctx = canvas.getContext('2d');

    // Draw template
    ctx.drawImage(templateImg, 0, 0, BILL_WIDTH, BILL_HEIGHT);

    // ------------------------------------------------------------------
    // 0. Redraw "VIRES IN NUMERIS" and year on the banner
    // ------------------------------------------------------------------
    const sansFont = 'Arial, Helvetica, sans-serif';

    // Cover old text with banner color (don't overlap medallion)
    ctx.fillStyle = BANNER_COLOR;
    ctx.fillRect(1082, 305, 1225 - 1082, 339 - 305);
    ctx.fillRect(BANNER_RIGHT_BOX.x1, BANNER_RIGHT_BOX.y1,
        BANNER_RIGHT_BOX.x2 - BANNER_RIGHT_BOX.x1,
        BANNER_RIGHT_BOX.y2 - BANNER_RIGHT_BOX.y1);

    // Left segment: "VIRES IN NUMERIS"
    {
        const lx1 = 1082 + 4, ly1 = 301;
        const lw = 1225 - 2 - lx1, lh = 343 - ly1;
        const motto = 'VIRES IN NUMERIS';
        ctx.fillStyle = BANNER_TEXT_COLOR;
        const fit = _fitFontToBox(ctx, motto, lw, lh, 24, sansFont);
        ctx.font = `bold ${fit.size}px ${sansFont}`;
        const metrics = ctx.measureText(motto);
        const tx = lx1 + (lw - metrics.width) / 2;
        const ty = ly1 + (lh + fit.size * 0.8) / 2;
        ctx.fillText(motto, tx, ty);
    }

    // Right segment: current year
    {
        const rx1 = BANNER_RIGHT_BOX.x1, ry1 = BANNER_RIGHT_BOX.y1;
        const rw = BANNER_RIGHT_BOX.x2 - rx1, rh = BANNER_RIGHT_BOX.y2 - ry1;
        const yearStr = new Date().getUTCFullYear().toString();
        ctx.fillStyle = BANNER_TEXT_COLOR;
        const fit = _fitFontToBox(ctx, yearStr, rw, rh, 24, sansFont);
        ctx.font = `bold ${fit.size}px ${sansFont}`;
        const metrics = ctx.measureText(yearStr);
        const tx = rx1 + (rw - metrics.width) / 2;
        const ty = ry1 + (rh + fit.size * 0.8) / 2;
        ctx.fillText(yearStr, tx, ty);
    }

    // ------------------------------------------------------------------
    // 1. Address QR code — fill the left white box (284×284)
    // ------------------------------------------------------------------
    {
        const { x1, y1, x2, y2 } = ADDRESS_QR_BOX;
        const boxW = x2 - x1;
        const boxH = y2 - y1;
        const qrSide = Math.min(boxW, boxH);
        const matrix = QRGenerator.generateQR(address, QRGenerator.EC_M);
        const pasteX = x1 + Math.floor((boxW - qrSide) / 2);
        const pasteY = y1 + Math.floor((boxH - qrSide) / 2);
        const moduleSize = qrSide / (matrix.length + 4); // +4 for 2-module border each side
        QRGenerator.qrToCanvas(matrix, ctx, pasteX, pasteY, moduleSize, 2);
    }

    // ------------------------------------------------------------------
    // 2. Private Key QR code — fill the right white box (284×285)
    //    Encodes a sweep URL (with WIF, network, and type params) so
    //    scanning the QR opens the sweep page with fields pre-populated.
    // ------------------------------------------------------------------
    {
        const { x1, y1, x2, y2 } = PRIVKEY_QR_BOX;
        const boxW = x2 - x1;
        const boxH = y2 - y1;
        const qrSide = Math.min(boxW, boxH);

        // Build sweep URL for the QR code.
        // Use the current host so the QR works whether the site is served
        // from GitHub Pages, a local Python server, or any other origin.
        // Falls back to the GitHub Pages URL when window.location is unavailable.
        let sweepBase;
        if (typeof window !== 'undefined' && window.location) {
            sweepBase = new URL('sweep.html', window.location.href).href;
        } else {
            sweepBase = SWEEP_PAGE_URL;
        }
        const params = new URLSearchParams();
        params.set('wif', privateKeyWif);
        params.set('network', network);
        let sweepType = addressType || 'taproot';
        if (addressType === 'taproot' && isTweaked) sweepType = 'taproot_tweaked';
        params.set('type', sweepType);
        const qrContent = sweepBase + '?' + params.toString();

        const matrix = QRGenerator.generateQR(qrContent, QRGenerator.EC_M);
        const pasteX = x1 + Math.floor((boxW - qrSide) / 2);
        const pasteY = y1 + Math.floor((boxH - qrSide) / 2);
        const moduleSize = qrSide / (matrix.length + 4);
        QRGenerator.qrToCanvas(matrix, ctx, pasteX, pasteY, moduleSize, 2);
    }

    // ------------------------------------------------------------------
    // 2b. "(tweaked)" label inside private key QR white box
    // ------------------------------------------------------------------
    if (isTweaked) {
        const tweakedLabel = '(tweaked)';
        const fontSize = 12;
        ctx.font = `${fontSize}px ${sansFont}`;
        ctx.fillStyle = 'rgb(30, 30, 30)';
        const metrics = ctx.measureText(tweakedLabel);
        const tx = PRIVKEY_QR_BOX.x2 - metrics.width - 3;
        const ty = PRIVKEY_QR_BOX.y2 - 3;
        ctx.fillText(tweakedLabel, tx, ty);
    }

    // ------------------------------------------------------------------
    // 3. Private key text — orange strip above the dollar box and QR
    // ------------------------------------------------------------------
    {
        const { x1, y1, x2, y2 } = PRIVKEY_TEXT_AREA;
        const tw = x2 - x1;
        const th = y2 - y1;
        const monoFont = '"Courier New", monospace';

        ctx.fillStyle = 'rgb(30, 30, 30)';

        if (isTweaked) {
            const suffix = ' (tweaked)';
            const suffixFont = `12px ${sansFont}`;
            ctx.font = suffixFont;
            const suffixW = ctx.measureText(suffix).width;

            const wifAvailW = tw - suffixW - 2;
            const fit = _fitFontToBox(ctx, privateKeyWif, wifAvailW, th, 24, monoFont);

            // Right-align: suffix flush to x=1808, WIF just before
            const suffixX = 1808 - suffixW;
            const textX = suffixX - fit.width - 2;
            const textY = y1 + (th + fit.size * 0.8) / 2;

            ctx.font = fit.font;
            ctx.fillText(privateKeyWif, textX, textY);

            ctx.font = suffixFont;
            const suffixY = y1 + (th + 12 * 0.8) / 2;
            ctx.fillText(suffix, suffixX, suffixY);
        } else {
            const fit = _fitFontToBox(ctx, privateKeyWif, tw, th, 24, monoFont);
            // Right-align to x=1808
            const textX = 1808 - fit.width;
            const textY = y1 + (th + fit.size * 0.8) / 2;
            ctx.font = fit.font;
            ctx.fillText(privateKeyWif, textX, textY);
        }
    }

    // ------------------------------------------------------------------
    // 4. Address text — bottom white band (801×58)
    //    Use condensed font for Taproot (longer addresses)
    // ------------------------------------------------------------------
    {
        const { x1, y1, x2, y2 } = ADDRESS_TEXT_BOX;
        const tw = x2 - x1;
        const th = y2 - y1;
        // Use condensed for taproot, regular mono for segwit
        const addrFontFamily = addressType === 'taproot'
            ? '"Arial Narrow", "Helvetica Neue", Arial, sans-serif'
            : '"Courier New", monospace';
        const fit = _fitFontToBox(ctx, address, tw, th, 36, addrFontFamily);

        ctx.fillStyle = 'rgb(30, 30, 30)';
        ctx.font = fit.font;
        const textX = x1 + (tw - fit.width) / 2;
        const textY = y1 + (th + fit.size * 0.8) / 2;
        ctx.fillText(address, textX, textY);
    }

    // ------------------------------------------------------------------
    // 5. Timestamp — vertical text at bottom-right corner
    // ------------------------------------------------------------------
    {
        const now = new Date();
        const timestamp = now.getUTCFullYear() + '-' +
            String(now.getUTCMonth() + 1).padStart(2, '0') + '-' +
            String(now.getUTCDate()).padStart(2, '0') + ' ' +
            String(now.getUTCHours()).padStart(2, '0') + ':' +
            String(now.getUTCMinutes()).padStart(2, '0') + ':' +
            String(now.getUTCSeconds()).padStart(2, '0') + ' UTC';

        const tsFontSize = 14;
        ctx.font = `${tsFontSize}px "Courier New", monospace`;
        ctx.fillStyle = 'rgb(0, 0, 0)';

        const tsMetrics = ctx.measureText(timestamp);
        const tsW = tsMetrics.width;

        // Rotate 90° counter-clockwise for vertical text
        ctx.save();
        const edgeMargin = 8;
        const px = BILL_WIDTH - tsFontSize - edgeMargin;
        const py = BILL_HEIGHT - edgeMargin;
        ctx.translate(px, py);
        ctx.rotate(-Math.PI / 2);
        ctx.fillText(timestamp, 0, tsFontSize * 0.8);
        ctx.restore();
    }
}

/**
 * Generate a bill and return it as a data URL (PNG).
 *
 * @param {HTMLImageElement} templateImg - Pre-loaded bill template image
 * @param {string} address - Bitcoin address
 * @param {string} privateKeyWif - WIF-encoded private key
 * @param {string} addressType - "segwit" or "taproot"
 * @param {boolean} isTweaked - If true, add "(tweaked)" labels
 * @param {string} network - "mainnet", "testnet4", or "regtest"
 * @returns {string} Data URL of the bill image (PNG)
 */
function generateBillDataURL(templateImg, address, privateKeyWif, addressType, isTweaked, network) {
    const canvas = document.createElement('canvas');
    generateBillOnCanvas(canvas, templateImg, address, privateKeyWif, addressType, isTweaked, network);
    return canvas.toDataURL('image/png');
}

/**
 * Pre-load the bill template image.
 *
 * @param {string} templatePath - URL or path to bill_template.png
 * @returns {Promise<HTMLImageElement>} Loaded image element
 */
function loadBillTemplate(templatePath) {
    if (templatePath === undefined) templatePath = 'bill_template.png';
    return new Promise((resolve, reject) => {
        const img = new Image();
        img.onload = () => resolve(img);
        img.onerror = () => reject(new Error('Failed to load bill template: ' + templatePath));
        img.src = templatePath;
    });
}


// ============================================================
// Exports
// ============================================================

const BillGenerator = {
    generateBillOnCanvas,
    generateBillDataURL,
    loadBillTemplate,
    BILL_WIDTH,
    BILL_HEIGHT,
};

if (typeof window !== 'undefined') {
    window.BillGenerator = BillGenerator;
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = BillGenerator;
}

File: server/bill_generator.py
"""
Bill image generator.
Takes the template bill image and overlays Bitcoin address/key text and QR codes.
"""

import io
import os
import base64
from datetime import datetime, timezone
from PIL import Image, ImageDraw, ImageFont
from qr_generator import generate_qr_image, EC_M, EC_Q

# Template bill image path (resolve relative to project root)
_DIR = os.path.dirname(os.path.abspath(__file__))
_PROJECT_ROOT = os.path.dirname(_DIR)
TEMPLATE_PATH = os.path.join(_PROJECT_ROOT, "assets", "bill_template.png")

# ============================================================
# Overlay positions (measured from pixel-level scan of the template)
# Bill image: 1843 x 784 pixels
#
# There are FOUR white regions on the template:
#   1. Left QR box:       x=35..319,   y=469..752  (284x284)  — Address QR
#   2. Right QR box:      x=1525..1808, y=40..324   (284x285)  — Private Key QR
#   3. Dollar-sign box:   x=1167..1494, y=40..107    (328x68)   — LEFT EMPTY (handwritten)
#   4. Bottom text band:  x=348..959,  y=700..753   (612x54)   — Address text
#   5. Orange strip above boxes: y≈5..38, x=1100..1808          — Private Key text
#
# Yellow "VIRES IN NUMERIS" banner with blue text:
#   Left segment (text):  x=1082..1227, y=305..338  — "VIRES IN NUMERIS"
#   Medallion gap:        x≈1228..1325
#   Right segment (year): x=1326..1424, y=305..338  — "2023" (to be replaced)
#   Banner background:    rgb(253, 229, 167)
#   Text color:           rgb(0, 161, 210)
# ============================================================

ADDRESS_QR_BOX    = (35,   469, 319,  752)   # 284×284 px
PRIVKEY_QR_BOX    = (1525,  40, 1808, 324)   # 284×285 px
PRIVKEY_TEXT_AREA  = (1100,   2, 1808,  30)   # Orange strip, closer to top edge
ADDRESS_TEXT_BOX  = (348,  694, 1148, 751)   # 801×58 px, shifted 2px up

# Banner properties
BANNER_COLOR       = (253, 229, 167)
BANNER_TEXT_COLOR   = (0, 161, 210)
BANNER_LEFT_BOX    = (1082, 305, 1228, 339)   # "VIRES IN NUMERIS" area (full extent)
BANNER_RIGHT_BOX   = (1326, 305, 1425, 339)   # year area


# ============================================================
# Font helpers
# ============================================================

_MONO_FONT_PATHS = [
    # Linux
    "/usr/share/fonts/truetype/dejavu/DejaVuSansMono-Bold.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf",
    "/usr/share/fonts/truetype/liberation/LiberationMono-Regular.ttf",
    "/usr/share/fonts/truetype/ubuntu/UbuntuMono-R.ttf",
    # macOS
    "/System/Library/Fonts/Menlo.ttc",
    "/System/Library/Fonts/SFMono-Regular.otf",
    "/Library/Fonts/Courier New.ttf",
    # Windows
    "C:/Windows/Fonts/consola.ttf",
    "C:/Windows/Fonts/cour.ttf",
]

_SANS_FONT_PATHS = [
    # Linux (bold first)
    "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf",
    "/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
    "/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf",
    # macOS
    "/System/Library/Fonts/Helvetica-Bold.ttc",
    "/System/Library/Fonts/Helvetica.ttc",
    "/System/Library/Fonts/SFNSText.ttf",
    # Windows
    "C:/Windows/Fonts/arialbd.ttf",
    "C:/Windows/Fonts/arial.ttf",
    "C:/Windows/Fonts/calibrib.ttf",
    "C:/Windows/Fonts/calibri.ttf",
]

# Narrow / condensed font for long addresses (e.g. Taproot 62-char)
# so they fill the text box with a similar visual weight to shorter addresses.
_NARROW_FONT_PATHS = [
    # Linux
    "/usr/share/fonts/truetype/liberation/LiberationSansNarrow-Regular.ttf",
    "/usr/share/fonts/truetype/liberation/LiberationSansNarrow-Bold.ttf",
    # macOS
    "/System/Library/Fonts/HelveticaNeue-Condensed.ttc",
    "/System/Library/Fonts/Supplemental/Arial Narrow.ttf",
    # Windows
    "C:/Windows/Fonts/arialn.ttf",
    "C:/Windows/Fonts/arialnb.ttf",
]


def _load_font(size, font_paths=None):
    """Load a font at the requested point size, with fallback."""
    if font_paths is None:
        font_paths = _MONO_FONT_PATHS
    for fp in font_paths:
        try:
            return ImageFont.truetype(fp, size)
        except (OSError, IOError):
            continue
    return ImageFont.load_default()


def _fit_font_to_box(draw, text, box_w, box_h, start_size=36, font_paths=None):
    """Find the largest font size where `text` fits within box_w x box_h."""
    for size in range(start_size, 5, -1):
        font = _load_font(size, font_paths)
        bbox = draw.textbbox((0, 0), text, font=font)
        tw = bbox[2] - bbox[0]
        th = bbox[3] - bbox[1]
        if tw <= box_w and th <= box_h:
            return font, tw, th
    font = _load_font(6, font_paths)
    bbox = draw.textbbox((0, 0), text, font=font)
    return font, bbox[2] - bbox[0], bbox[3] - bbox[1]


# ============================================================
# Main generator
# ============================================================

def generate_bill_image(address, private_key_wif, address_type="taproot",
                        is_tweaked=False, sweep_url=None):
    """Generate a bill image with the given address and private key.

    Args:
        address: Bitcoin address string (bc1q... or bc1p...)
        private_key_wif: WIF-encoded private key
        address_type: "segwit" or "taproot" for display purposes
        is_tweaked: If True, render "(tweaked)" labels next to the private
                    key text and below the private key QR code.  Used for
                    Taproot addresses that have a backup key (script tree).
        sweep_url: If provided, the private key QR code encodes this URL
                   instead of the raw WIF (e.g. sweep page URL with WIF
                   and network query parameters).

    Returns:
        PIL Image of the completed bill
    """
    template = Image.open(TEMPLATE_PATH).convert("RGB")
    draw = ImageDraw.Draw(template)

    # ------------------------------------------------------------------
    # 0. Redraw "VIRES IN NUMERIS" and year on the banner
    #    Instead of covering with a rectangle (which would overlap the
    #    medallion circle), replace only the old blue text pixels with
    #    the banner background color, then redraw with the new font.
    # ------------------------------------------------------------------
    # Cover with solid rectangles — sized to NOT overlap the medallion.
    # Banner: x=1082..1225, medallion starts at ~x=1226.
    draw.rectangle((1082, 305, 1225, 339), fill=BANNER_COLOR)
    draw.rectangle(BANNER_RIGHT_BOX, fill=BANNER_COLOR)

    # Left segment: "VIRES IN NUMERIS"
    # Banner spans x=1082..1225 (144px). Add equal margins on both sides
    # so the text is centered between the banner left edge and medallion.
    lx1, ly1 = 1082 + 4, 301
    lx2_text, ly2 = 1225 - 2, 343
    lw, lh = lx2_text - lx1, ly2 - ly1
    motto = "VIRES IN NUMERIS"
    font, tw, th = _fit_font_to_box(draw, motto, lw, lh,
                                     start_size=24, font_paths=_SANS_FONT_PATHS)
    draw.text((lx1 + (lw - tw) // 2, ly1 + (lh - th) // 2),
              motto, fill=BANNER_TEXT_COLOR, font=font)

    # Right segment: current year — fit to the right box independently
    # so it appears visually the same weight/height as the motto text.
    rx1, ry1, rx2, ry2 = BANNER_RIGHT_BOX
    rw, rh = rx2 - rx1, ry2 - ry1
    year_str = str(datetime.now(timezone.utc).year)
    yr_font, tw, th = _fit_font_to_box(draw, year_str, rw, rh,
                                        start_size=24, font_paths=_SANS_FONT_PATHS)
    draw.text((rx1 + (rw - tw) // 2, ry1 + (rh - th) // 2),
              year_str, fill=BANNER_TEXT_COLOR, font=yr_font)

    # ------------------------------------------------------------------
    # 1. Address QR code — fill the left white box (284×284)
    # ------------------------------------------------------------------
    ax1, ay1, ax2, ay2 = ADDRESS_QR_BOX
    box_w = ax2 - ax1
    box_h = ay2 - ay1
    qr_side = min(box_w, box_h)
    addr_qr = generate_qr_image(address, size=qr_side, ec_level=EC_M)
    # Centre the square QR in the box
    paste_x = ax1 + (box_w - qr_side) // 2
    paste_y = ay1 + (box_h - qr_side) // 2
    template.paste(addr_qr, (paste_x, paste_y))

    # ------------------------------------------------------------------
    # 2. Private Key QR code — fill the right white box (284×285)
    #    Keep it square using the smaller dimension, centred in the box.
    #    If sweep_url is provided, encode the URL instead of the raw WIF.
    # ------------------------------------------------------------------
    px1, py1, px2, py2 = PRIVKEY_QR_BOX
    box_w = px2 - px1
    box_h = py2 - py1
    qr_side = min(box_w, box_h)
    qr_content = sweep_url if sweep_url else private_key_wif
    priv_qr = generate_qr_image(qr_content, size=qr_side, ec_level=EC_M)
    paste_x = px1 + (box_w - qr_side) // 2
    paste_y = py1 + (box_h - qr_side) // 2
    template.paste(priv_qr, (paste_x, paste_y))

    # ------------------------------------------------------------------
    # 2b. "(tweaked)" label inside the private key QR white box,
    #     bottom-right corner (overlaps the QR quiet zone).
    # ------------------------------------------------------------------
    if is_tweaked:
        tweaked_label = "(tweaked)"
        tw_font = _load_font(12, _SANS_FONT_PATHS)
        tw_bbox = draw.textbbox((0, 0), tweaked_label, font=tw_font)
        tw_w = tw_bbox[2] - tw_bbox[0]
        tw_h = tw_bbox[3] - tw_bbox[1]
        # Bottom-right of the QR white box with small margin
        tw_x = px2 - tw_w - 3
        tw_y = py2 - tw_h - 3
        draw.text((tw_x, tw_y), tweaked_label, fill=(30, 30, 30), font=tw_font)

    # ------------------------------------------------------------------
    # 3. Private key text — orange strip ABOVE the dollar box and QR
    #    (the dollar-sign box is left empty for handwriting)
    #    When is_tweaked, append " (tweaked)" to the right of the WIF.
    # ------------------------------------------------------------------
    tx1, ty1, tx2, ty2 = PRIVKEY_TEXT_AREA
    tw = tx2 - tx1   # ~708
    th = ty2 - ty1   # ~33

    if is_tweaked:
        # Measure the "(tweaked)" suffix in a smaller font
        suffix = " (tweaked)"
        suffix_font = _load_font(12, _SANS_FONT_PATHS)
        sb = draw.textbbox((0, 0), suffix, font=suffix_font)
        suffix_w = sb[2] - sb[0]
        suffix_h = sb[3] - sb[1]
        # Fit the WIF into a reduced-width area
        wif_avail_w = tw - suffix_w - 2
        font, text_w, text_h = _fit_font_to_box(
            draw, private_key_wif, wif_avail_w, th, start_size=24)
        # Right-align: suffix flush to x=1808, WIF just before it
        suffix_x = 1808 - suffix_w
        text_x = suffix_x - text_w - 2
        text_y = ty1 + (th - text_h) // 2
        draw.text((text_x, text_y), private_key_wif,
                  fill=(30, 30, 30), font=font)
        # Vertically centre the suffix relative to the WIF text
        suffix_y = ty1 + (th - suffix_h) // 2
        draw.text((suffix_x, suffix_y), suffix,
                  fill=(30, 30, 30), font=suffix_font)
    else:
        font, text_w, text_h = _fit_font_to_box(
            draw, private_key_wif, tw, th, start_size=24)
        # Right-align to the right edge of the QR box (x=1808)
        text_x = 1808 - text_w
        text_y = ty1 + (th - text_h) // 2
        draw.text((text_x, text_y), private_key_wif,
                  fill=(30, 30, 30), font=font)

    # ------------------------------------------------------------------
    # 4. Address text — bottom white band (801×58)
    #    Taproot addresses are 62 chars vs SegWit's 42 chars.
    #    Use a narrow/condensed font for Taproot so the text fills the
    #    box with similar visual weight to SegWit.
    # ------------------------------------------------------------------
    tx1, ty1, tx2, ty2 = ADDRESS_TEXT_BOX
    tw = tx2 - tx1
    th = ty2 - ty1
    addr_font_paths = _NARROW_FONT_PATHS if address_type == "taproot" else None
    font, text_w, text_h = _fit_font_to_box(draw, address, tw, th,
                                              start_size=36, font_paths=addr_font_paths)
    text_x = tx1 + (tw - text_w) // 2
    text_y = ty1 + (th - text_h) // 2
    draw.text((text_x, text_y), address, fill=(30, 30, 30), font=font)

    # ------------------------------------------------------------------
    # 5. Timestamp — vertical text at the bottom-right corner
    #    Readable when the bill is flipped 90° to the left (rotated CCW).
    #    Rendered with generous padding so it doesn't clip at the edge.
    # ------------------------------------------------------------------
    timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
    ts_font = _load_font(14)
    # Measure the text
    ts_bbox = draw.textbbox((0, 0), timestamp, font=ts_font)
    ts_w = ts_bbox[2] - ts_bbox[0]
    ts_h = ts_bbox[3] - ts_bbox[1]
    # Render onto a transparent image with padding, then rotate
    pad = 6
    ts_img = Image.new("RGBA", (ts_w + pad * 2, ts_h + pad * 2), (0, 0, 0, 0))
    ts_draw = ImageDraw.Draw(ts_img)
    ts_draw.text((pad, pad), timestamp, fill=(0, 0, 0, 255), font=ts_font)
    # Rotate 90° counter-clockwise so it reads bottom-to-top
    ts_rotated = ts_img.rotate(90, expand=True)
    # Position closer to the right edge of the bill
    img_w, img_h = template.size
    edge_margin = 8
    paste_x = img_w - ts_rotated.width - edge_margin
    paste_y = img_h - ts_rotated.height - edge_margin
    template.paste(ts_rotated, (paste_x, paste_y), ts_rotated)

    return template


def bill_to_png_bytes(bill_image):
    """Convert a bill PIL Image to PNG bytes."""
    buf = io.BytesIO()
    bill_image.save(buf, format='PNG', quality=95)
    return buf.getvalue()


def bill_to_base64(bill_image):
    """Convert a bill PIL Image to base64-encoded PNG string."""
    png_bytes = bill_to_png_bytes(bill_image)
    return base64.b64encode(png_bytes).decode('ascii')


if __name__ == "__main__":
    from bitcoin_crypto import generate_segwit_address, generate_taproot_address

    sw = generate_segwit_address(mainnet=True)
    bill = generate_bill_image(sw["address"], sw["private_key_wif"], "segwit")
    bill.save("test_bill_segwit.png")
    print(f"SegWit bill: {sw['address']}")

    tr = generate_taproot_address(mainnet=True, backup_key=False)
    bill = generate_bill_image(tr["address"], tr["private_key_wif"], "taproot")
    bill.save("test_bill_taproot.png")
    print(f"Taproot bill (no backup): {tr['address']}")

    from bitcoin_crypto import private_key_to_wif
    tr2 = generate_taproot_address(mainnet=True, backup_key=True)
    tweaked_wif = private_key_to_wif(
        bytes.fromhex(tr2["tweaked_private_key_hex"]), mainnet=True)
    bill = generate_bill_image(tr2["address"], tweaked_wif, "taproot",
                               is_tweaked=True)
    bill.save("test_bill_taproot_tweaked.png")
    print(f"Taproot bill (tweaked/backup): {tr2['address']}")
